packages/core/src/browser/label-provider.ts

```diff
@@ -38,25 +38,31 @@ export interface LabelProviderContribution {
     /**
      * returns an icon class for the given element.
      */
-    getIcon?(element: object): MaybePromise<string>;
+    getIcon?(element: object): string | undefined;
 
     /**
      * returns a short name for the given element.
      */
-    getName?(element: object): string;
+    getName?(element: object): string | undefined;
 
     /**
      * returns a long name for the given element.
      */
-    getLongName?(element: object): string;
+    getLongName?(element: object): string | undefined;
 
     /**
      * Emit when something has changed that may result in this label provider returning a different
      * value for one or more properties (name, icon etc).
      */
     readonly onDidChange?: Event<DidChangeLabelEvent>;
 
-    readonly getConstituentUris?: (compositeElement: object) => URI[];
+    /**
+     * Check whether the given element is affected by the given change event.
+     * Contributions delegating to the label provider can use this hook
+     * to perfrom a recursive check.
+     */
+    affects?(element: object, event: DidChangeLabelEvent): boolean;
```

Comment:
@westbury I've removed dependencies to `URI` with such method

Reply:
That almost solves it.  However consider two URLs A and B combined into a diff {A,B} .  Now take a diff between that diff file and a third file C, so we have {{A,B},C}.  I know we should not get such a URI but as far as the label providers are concerned, this is supported.  Pass in the URI for {{A,B},C} together with an event that affects A only.  The implementation in DiffUris.affects will pass the URIs for {A,B} and C directly to the event and will not detect the change.

Reply:
Is there a real case? I can solve it by passing affects fn which does recursive check, but it introduces all kind of complexity, like what if `C` can be broke down to A and B again. How do stop recursion? I would prefer to avoid make it complex trying to solve imaginary issues.